### PR TITLE
fix bugs found using afl

### PR DESCRIPTION
**Pull request by ddcc**

### CODE REVIEW

File: src/wasm.h [hunk 1]
```diff
@@ -961,6 +961,7 @@ class If : public SpecificExpression<Expression::IfId> {
 
   void finalize() {
     if (ifFalse) {
+      assert(ifTrue && ifFalse);
```
Comment (kripken) [inline src/wasm.h]: First thing, thanks for this! :)

And a comment on this line: this should not be possible, ifTrue must always exist. So this assertion is valid, however, the problem is happening at an earlier stage, so we should also add another fix somewhere. Is it possible to get the input that led to this?

File: src/wasm.h [hunk 2]
```diff
@@ -1244,6 +1245,7 @@ class Binary : public SpecificExpression<Expression::BinaryId> {
     if (isRelational()) {
       type = i32;
     } else {
+      assert(left && right);
```
Comment (kripken) [inline src/wasm.h]: same question about this, and then next change as well.

**Comment by kripken** [comment]
Before we merge this in, have you joined the [w3c group](https://www.w3.org/community/webassembly/participants)?

**Comment by ddcc** [comment]
Yes, I've joined the working group through Google. I have the failing binary testcases; the earlier fix for the assertions in `finalize()` is the new assert at wasm-binary.c:1483.

File: src/wasm-binary.h [hunk 3]
```diff
@@ -1195,7 +1195,7 @@ class WasmBinaryBuilder {
   bool debug;
 
   size_t pos = 0;
-  int32_t startIndex = -1;
+  int64_t startIndex = -1;
```
Comment (kripken) [inline src/wasm-binary.h]: we read a 32-bit LEB into this, why isn't int32_t enough?

File: src/wasm.h (hunk 1 above)
Comment (kripken) [inline src/wasm.h]: this assert could be an `assert(ifTrue)` on the previous line.

File: src/wasm.h (hunk 2 above)
Comment (kripken) [inline src/wasm.h]: This assert could be before the `if` begins, as it is relevant for both cases.

File: src/wasm-binary.h (hunk 3 above)
Comment (ddcc) [inline src/wasm-binary.h]: Correct me if I'm interpreting the code wrong, but isn't it reading an unsigned 32-bit LEB into a signed 32-bit integer type, which will overflow the upper half of values to negative?

Comment (kripken) [inline src/wasm-binary.h]: Yes, this should be `uint32_t`, unsigned. But really, it should be the `Index` type.

(In theory it's also bad that we use `-1` as an invalid value, but the possibility of all 32 bits being used as function indexes is not practical. Might be worth a TODO note though.)

**Comment by ddcc** [comment]
Is there somewhere that I should put the malformed WASM testcases? The spec repository seems to only have WAST testcases.

**Comment by kripken** [comment]
Correct. We could add support for binary tests here, however, are these perhaps test that make sense to add to the spec repo? I'm not sure. I guess if they are highly specific to bugs in this project, perhaps not, but if they test general things like "not enough functions provided for what is declared earlier" then that does seem like a spec repo thing.

**Comment by ddcc** [comment]
They are probably generic enough to be put in the spec repo. I did some fuzzing on the interpreter from WebAssembly/sexpr-wasm-prototype, and there were some similar problems. I can also start writing up some tests for the issues from #549.

**Comment by kripken** [comment]
Ok, great, let's merge this in.

**Pull request merged**